deepspeed/runtime/checkpoint_engine/checkpoint_engine.py

```diff
@@ -0,0 +1,23 @@
+import torch
+from deepspeed.utils import logger
+
+
+class CheckpointEngine(object):
+    def __init__(self):
+        pass
+
+    def save(self, state_dict, path: str):
+        logger.info(f"Saving {path}...")
+        torch.save(state_dict, path)
+        logger.info(f"Saved {path}.")
+        return None
+
+    def load(self, path: str, map_location=None):
+        logger.info(f"Loading checkpoint from {path}...")
+        partition = torch.load(path, map_location=map_location)
+        logger.info(f"Loaded checkpoint from {path}.")
+        return partition
+
+    def commit(self, tag):
```

Comment:
It plays the role to tell main process current checkpoint is saved completely. For raw torch, it will be same as following code.
```python
        if save_latest and self.global_rank == 0:
            with open(os.path.join(save_dir, 'latest'), 'w') as fd:
                fd.write(tag)
```

Nebula need this signal to start the persistence processing for checkpoint.

Reply:
![image](https://user-images.githubusercontent.com/13343117/180011966-c8763665-9d5a-4b1b-a569-20c5c9d1b382.png)
Actually for nebula.save, we have the following steps to persist a checkpoint into disk:
1. **tier1**: we pin the shared memory and write the checkpoints files into shared memory(let say under /dev/shm/). It is a synced way which will block the training.
2. **tier3**: when nebule service received the `commit` signal from SDK, nebula service will start a persistent process which move the checkpoints from tier1 to tier3(equal to `nebula_load_path_tier3` where you commented). After tier3 done, the nebula meta info will be wrote into tier3 path.

So for your questions:
> For nebula, do we need a single commit for each checkpoint file or just one for the save_checkpoint() call?
    Nebula needs every training process to call `commit` to tell the services current checkpoint can be moved to tier3. But origin torch only write the meta file for checkpoint management.

>  what does the return value mean?
Emmm, it is dummy return for torch I think. If needed, maybe we can move the torch meta file operation into this method only for rank0? What do you think? :)

Reply:
One thing I want to call out is that: the files structures are totally different for original torch path and nebula tier3 path.

Reply:
It is fine for the checkpoint file structure to be differerent between torch and nebula tier3, it simply means that `torch.load()` cannot be used for nebula checkpoints. This is not a concern for deepspeed. 

I think the `fd.write(tag)` should remain inside deepspeed and outside nebula for a number of reasons: (1) it is not currently created by torch since it is not checkpoint state, but metadata, (2) It is used by deepspeed and existing clients for checkpoint management, and (3) it is very small data and fast to write. Typically, it is used to identify the latest checkpoint file in a folder of many checkpoint files, but clients can also use tags in other ways. So, can you explain why this part has to change for nebula?

Reply:
It makes sense to keep `fd.write(tag)` out of nebula. Actually we did that in code as well.
```python
# Save latest checkpoint tag
dist.barrier()
self.checkpoint_engine.commit(tag)
if save_latest and self.global_rank == 0:
    with open(os.path.join(save_dir, 'latest'), 'w') as fd:
        fd.write(tag)
```
Then for:
1. `torch`: the `commit` was used to align the API. No extra efforts for condition(`if nebula else...`).
2. `nebula`: the `commit` will send the signal to nebula services to start tier3.

The `fd.write(tag)` is workable for torch, but cannot replace `commit()` to tell nebula service current checkpoint is saved.

So for you question:  
> why this part has to change for nebula?
We need the `commit()` for nebula when all files under a tag are saved. But it will just return dummy true for `torch`.

Reply:
> It makes sense to keep `fd.write(tag)` out of nebula. Actually we did that in code as well.

Sorry, I misread the PR. You are correct that you did not change this code.

Reply:
> Then for:
> 
> 1. `torch`: the `commit` was used to align the API. No extra efforts for condition(`if nebula else...`).
> 2. `nebula`: the `commit` will send the signal to nebula services to start tier3.
> 
> The `fd.write(tag)` is workable for torch, but cannot replace `commit()` to tell nebula service current checkpoint is saved.
> 
> So for you question:
> 
> > why this part has to change for nebula?
> > We need the `commit()` for nebula when all files under a tag are saved. But it will just return dummy true for `torch`.

So, this brings up the importance of clarifying the semantics of nebula and other checkpoint engines. I don't know if there is standard terminology for these concepts. But it seems to me that one key question is whether persistence is implicit or explicit. Persistence with `torch.save()` is implicit because checkpoint files are persisted on disk on return, while nebula is explicit because an additional call is required to make it persist the files. I think this is an important consideration for designing the general CheckpointEngine abstraction. We can discuss this further offline.